xds/src/main/java/io/grpc/xds/ClusterManagerLoadBalancer.java

```diff
@@ -0,0 +1,264 @@
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.internal.ServiceConfigUtil.PolicySelection;
+import io.grpc.util.ForwardingLoadBalancerHelper;
+import io.grpc.util.GracefulSwitchLoadBalancer;
+import io.grpc.xds.ClusterManagerLoadBalancerProvider.ClusterManagerConfig;
+import io.grpc.xds.XdsLogger.XdsLogLevel;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * The top-level load balancing policy.
+ */
+class ClusterManagerLoadBalancer extends LoadBalancer {
+
+  @VisibleForTesting
+  static final int DELAYED_ACTION_DELETION_TIME_MINUTES = 15;
+  @VisibleForTesting
+  static final CallOptions.Key<String> ROUTING_CLUSTER_NAME_KEY =
+      CallOptions.Key.create("io.grpc.xds.ROUTING_CLUSTER_NAME_KEY");
+
+  private final Map<String, ChildLbState> childLbStates = new HashMap<>();
+  private final Helper helper;
+  private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService timeService;
+  private final XdsLogger logger;
+
+  ClusterManagerLoadBalancer(Helper helper) {
+    this.helper = checkNotNull(helper, "helper");
+    this.syncContext = checkNotNull(helper.getSynchronizationContext(), "syncContext");
+    this.timeService = checkNotNull(helper.getScheduledExecutorService(), "timeService");
+    logger = XdsLogger.withLogId(InternalLogId.allocate("cds-lb", helper.getAuthority()));
+    logger.log(XdsLogLevel.INFO, "Created");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    logger.log(XdsLogLevel.DEBUG, "Received resolution result: {0}", resolvedAddresses);
+    ClusterManagerConfig config = (ClusterManagerConfig)
+        resolvedAddresses.getLoadBalancingPolicyConfig();
+    Map<String, PolicySelection> newChildPolicies = config.childPolicies;
+    logger.log(
+        XdsLogLevel.INFO,
+        "Received cluster_manager lb config: child names={0}", newChildPolicies.keySet());
+    for (Map.Entry<String, PolicySelection> entry : newChildPolicies.entrySet()) {
+      final String name = entry.getKey();
+      LoadBalancerProvider childPolicyProvider = entry.getValue().getProvider();
+      Object childConfig = entry.getValue().getConfig();
+      if (!childLbStates.containsKey(name)) {
+        childLbStates.put(name, new ChildLbState(name, childPolicyProvider));
+      } else {
+        childLbStates.get(name).reactivate(childPolicyProvider);
+      }
+      final ResolvedAddresses childAddresses =
+          resolvedAddresses.toBuilder().setLoadBalancingPolicyConfig(childConfig).build();
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          childLbStates.get(name).lb.handleResolvedAddresses(childAddresses);
```

Comment:
Declare variable for `childLbStates.get(name)` ahead of `syncContext.execute()`, otherwise `childLbStates` might have been mutated when `syncContext.execute() in extreme case.

Or just run `childLbStates.get(name).lb.handleResolvedAddresses(childAddresses)` inline at the end of this method.

Reply:
Good point. Thanks. Fixed.